skimage/filters/rank/core_cy.pyx

```diff
@@ -271,3 +286,293 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
         free(se_s_r)
         free(se_s_c)
         free(histo)
+
+
+cdef void _core_3D(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+                               dtype_t, Py_ssize_t, Py_ssize_t, double,
+                               double, Py_ssize_t, Py_ssize_t) nogil,
+                   dtype_t[:, :, ::1] image,
+                   char[:, :, ::1] selem,
+                   char[:, :, ::1] mask,
+                   dtype_t_out[:, :, :, ::1] out,
+                   signed char shift_x, signed char shift_y,
+                   signed char shift_z, double p0, double p1,
+                   Py_ssize_t s0, Py_ssize_t s1,
+                   Py_ssize_t n_bins) except *:
+    """Compute histogram for each pixel neighborhood, apply kernel function and
+    use kernel function return value for output image.
+    """
+
+    cdef Py_ssize_t planes = image.shape[0]
+    cdef Py_ssize_t rows = image.shape[1]
+    cdef Py_ssize_t cols = image.shape[2]
+    cdef Py_ssize_t splanes = selem.shape[0]
+    cdef Py_ssize_t srows = selem.shape[1]
+    cdef Py_ssize_t scols = selem.shape[2]
+    cdef Py_ssize_t odepth = out.shape[3]
+
+    cdef Py_ssize_t centre_p = <Py_ssize_t>(selem.shape[0] / 2) + shift_x
+    cdef Py_ssize_t centre_r = <Py_ssize_t>(selem.shape[1] / 2) + shift_y
+    cdef Py_ssize_t centre_c = <Py_ssize_t>(selem.shape[2] / 2) + shift_z
+
+    # check that structuring element center is inside the element bounding box
+    assert centre_r >= 0
+    assert centre_c >= 0
+    assert centre_p >= 0
+    assert centre_r < srows
+    assert centre_c < scols
+    assert centre_p < splanes
+
+    cdef Py_ssize_t mid_bin = n_bins / 2
+
+    # define pointers to the data
+    cdef char* mask_data = &mask[0, 0, 0]
+
+    # define local variable types
+    cdef Py_ssize_t r, c, p, j, rr, cc, pp, value, local_max, i, even_row
+
+    # number of pixels actually inside the neighborhood (double)
+    cdef double pop = 0
+
+    # build attack and release borders by using difference along axis
+    t = np.dstack((selem, np.zeros((selem.shape[0], selem.shape[1], 1))))
+    cdef unsigned char[:, :, :] t_e = (np.diff(t, axis=2) < 0).view(np.uint8)
+
+    t = np.dstack((np.zeros((selem.shape[0], selem.shape[1], 1)), selem))
+    cdef unsigned char[:, :, :] t_w = (np.diff(t, axis=2) > 0).view(np.uint8)
+
+    t = np.hstack((selem, np.zeros((selem.shape[0], 1, selem.shape[2]))))
+    cdef unsigned char[:, :, :] t_s = (np.diff(t, axis=1) < 0).view(np.uint8)
+
+    t = np.hstack((np.zeros((selem.shape[0], 1, selem.shape[2])), selem))
+    cdef unsigned char[:, :, :] t_n = (np.diff(t, axis=1) > 0).view(np.uint8)
```

Comment:
I can look into this but maybe @ryanlu41 could give a better answer?

Reply:
From my interpretation, which I figured out mostly after I wrote this, each plane gets its own rolling histogram, but this histogram is built from scratch for each plane. That is, if you had an image shaped (1000, 5, 5), this algorithm would perform quite poorly. However, most images have reasonably-sized planes, so you amortise the cost of building a rolling histogram after each plane, rather than having a rolling histogram across all axes. So, I think we can leave this for now, but @ryanlu41 let us know if you disagree or if my interpretation is totally incorrect.

Actually, if it *is* correct, it is worth adding a comment at this point that explains this.

Reply:
Yes, each plane gets its own rolling histogram. I would be happy to optimize it further, I just wanted to get a working implementation for now

Reply:
Let’s get a working implementation in first, and worry about optimising later.